### PR TITLE
Implement email domain filter

### PR DESCRIPTION
**Pull request by nisargjhaveri**
resolves T89

### CODE REVIEW

File: sql/users.sql [hunk 1]
```diff
@@ -60,10 +71,12 @@ CREATE TABLE IF NOT EXISTS `users` (
   `country` varchar(255) COLLATE utf8_bin NOT NULL,
   `dob` date NOT NULL,
   `organization` varchar(255) COLLATE utf8_bin NOT NULL,
+  `mail_hash` varchar(255) COLLATE utf8_bin NOT NULL,
```
Comment (nisargjhaveri) [inline sql/users.sql]: I added this for future use!

Comment (nisargjhaveri) [inline sql/users.sql]: Don't update production DB with this change.

File: src/app/controllers/auth.php [hunk 2]
```diff
@@ -374,6 +389,8 @@ private function handle_registration_by_email($action) {
                 $verified = $this->auth_model->verify_mail($hash, false);
                 if (!$verified) {
                     $error[] = "Invalid request";
+                } elseif(strlen($password) < 6) {
+                    $error[] = "Password must be at least 6 character long";
```
Comment (polybuildr) [inline src/app/controllers/auth.php]: "6 characters"

File: src/app/models/auth_model.php [hunk 3]
```diff
@@ -142,14 +144,16 @@ function create_user($oauth_id, $user_data) {
         $email_verified = isset($user_data["email_verified"]) ? $user_data["email_verified"] : "0";
         $resitration_status = isset($user_data["resitration_status"]) ? $user_data["resitration_status"] : "email_required";
```
Comment (polybuildr) [inline src/app/models/auth_model.php:145]: `resitration_status` :'(

File: src/app/models/auth_model.php [hunk 4]
```diff
@@ -301,4 +310,41 @@ function reset_ldap_password($email, $password) {
...
         ]);
     }
+
+    function is_good_email($email) {
+        $email_info = explode("@", $email);
+        $domain = end($email_info);
+
+        $stmt = $this->db_lib->prepared_execute(
+            $this->DB->users,
+            "SELECT `doamin` FROM `bad_domains` WHERE `doamin`=?",
```
Comment (polybuildr) [inline src/app/models/auth_model.php]: `doamin` ...

File: src/app/models/auth_model.php [hunk 5]
```diff
@@ -301,4 +310,41 @@ function reset_ldap_password($email, $password) {
...
+        }
+
+        $email_info = explode("@", $email);
+        $name = implode("@", array_slice($email_info, 0, -1));
+        $domain = end($email_info);
+
+        $name = strtolower($name);
+        $domain = strtolower($domain);
+
+        // All domain specigic rules goes here
```
Comment (polybuildr) [inline src/app/models/auth_model.php]: "specigic" ?

File: src/app/models/auth_model.php [hunk 6]
```diff
@@ -301,4 +310,41 @@ function reset_ldap_password($email, $password) {
...
+        if ($domain == "gmail.com") {
```
Comment (polybuildr) [inline src/app/models/auth_model.php:346]: Stuff like `googlemail.com`?

File: src/app/models/auth_model.php (hunk 4 above)
Comment (nisargjhaveri) [inline src/app/models/auth_model.php]: :'( :'( :'(

File: src/app/models/auth_model.php [hunk 7]
```diff
@@ -301,4 +310,41 @@ function reset_ldap_password($email, $password) {
...
+            $name = str_replace('.', '', explode('+', $name)[0]);
+        }
+
+        return $name . '@' . $domain;
```
Comment (polybuildr) [inline src/app/models/auth_model.php:350]: The email hash is just name@domain?

Comment (nisargjhaveri) [inline src/app/models/auth_model.php:350]: yes.

**Comment by polybuildr** [comment]
I looked through the code, it looks sane. Has it been tested?

File: src/app/models/auth_model.php (hunk 6 above)
Comment (nisargjhaveri) [inline src/app/models/auth_model.php:346]: What is googlemail.com!?

File: src/app/models/auth_model.php (hunk 7 above)
Comment (polybuildr) [inline src/app/models/auth_model.php:350]: Why is it called hash then?

**Comment by nisargjhaveri** [comment]
I tested it while writing.

File: src/app/models/auth_model.php (hunk 6 above)
Comment (polybuildr) [inline src/app/models/auth_model.php:346]: Try sending an email to `your_username@googlemail.com`.

Comment (nisargjhaveri) [inline src/app/models/auth_model.php:346]: Why you do this Google. :/

Comment (polybuildr) [inline src/app/models/auth_model.php:346]: xP

**Comment by nisargjhaveri** [comment]
Updated.

File: sql/users.sql [hunk 8]
```diff
@@ -17,6 +17,17 @@ SET time_zone = "+00:00";
 -- --------------------------------------------------------
 
 --
+-- Table structure for table `bad_domains`
+--
+
+CREATE TABLE IF NOT EXISTS `bad_domains` (
+  `doamin` varchar(255) NOT NULL,
```
Comment (polybuildr) [inline sql/users.sql]: The schema still has `doamin`.

Comment (nisargjhaveri) [inline sql/users.sql]: Did you even check! :P

**Comment by polybuildr** [comment]
I read through, code looks sane. @hharchani will merge after testing.

**Comment by hharchani** [comment]
I've tested, works fine. Merging

**Pull request merged**